tests/posix/common/src/semaphore.c

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <errno.h>
+#include <pthread.h>
+#include <semaphore.h>
+
+#define STACK_SIZE 1024
+
+sem_t sema;
+void *dummy_sem;
+
+static K_THREAD_STACK_DEFINE(stack, STACK_SIZE);
+
+static void *child_func(void *p1)
+{
+	zassert_equal(sem_post(dummy_sem), -1, "sem_post of"
+		      " invalid semaphore object didn't fail");
+	zassert_equal(errno, EINVAL, NULL);
+
+	zassert_equal(sem_post(&sema), 0, "sem_post failed");
+	return NULL;
+}
+
+void test_posix_semaphore(void)
+{
+	pthread_t newthread;
+	pthread_attr_t attr;
+	struct sched_param schedparam;
+	int schedpolicy = SCHED_FIFO;
+	int val, ret;
+	struct timespec abstime;
+
+	schedparam.priority = 1;
+
+	ret = pthread_attr_init(&attr);
+	if (ret != 0) {
+		zassert_equal(pthread_attr_destroy(&attr), 0,
+			      "Unable to destroy pthread object attrib");
+		zassert_equal(pthread_attr_init(&attr), 0,
+			      "Unable to create pthread object attrib");
+	}
+
+	pthread_attr_setstack(&attr, &stack, STACK_SIZE);
+	pthread_attr_setschedpolicy(&attr, schedpolicy);
+	pthread_attr_setschedparam(&attr, &schedparam);
+
+	zassert_equal(sem_init(&sema, 0, (CONFIG_SEM_VALUE_MAX + 1)), -1,
+		      "value larger than %d\n", CONFIG_SEM_VALUE_MAX);
+	zassert_equal(errno, EINVAL, NULL);
+
+	zassert_equal(sem_init(&sema, 0, 0), 0, "sem_init failed");
+
+	zassert_equal(sem_getvalue(&sema, &val), 0, NULL);
+	zassert_equal(val, 0, NULL);
+
+	ret = pthread_create(&newthread, &attr, child_func, NULL);
+	zassert_equal(ret, 0, "Thread creation failed");
+
+	zassert_equal(clock_gettime(CLOCK_REALTIME, &abstime), 0,
+		      "clock_gettime failed");
+
+	abstime.tv_sec += 5;
+
+	zassert_equal(sem_timedwait(&sema, &abstime), 0, NULL);
+
+	zassert_equal(sem_timedwait(&sema, &abstime), -1, NULL);
+	zassert_equal(errno, ETIMEDOUT, NULL);
+
+	zassert_equal(sem_trywait(&sema), -1, NULL);
+	zassert_equal(errno, EAGAIN, NULL);
+
+	zassert_equal(sem_destroy(&sema), 0, "semaphore is not destroyed");
```

Comment:
You could replace lines 77 to 87 with a `sem_post(&sema)`: I mean to say, reuse both semaphore and pthread attributes without having to delete and initialize again - it has the same effect. You can observe this in your initializing thread attributes: you haven't set the `schedparam.priority` but it still works in creating the thread with same previous priority, as the thread resources are still available, and are only released with `pthread_join` or `pthread_detach`.

Reply:
But the child thread might have terminated already by the time we call sem_post. And trying to have synchronization without having threads to synchronize would logically may not make sense. Also schedparam.priority has nothing to do with pthread_attr_destroy. It is just a structure with one of its member's initialized.

Reply:
Okay, agreed on the logically having two threads for synchronizing - but can we have the same semaphore for doing it? Can `sched_yield`, make the child thread give the semaphore. Just check if deleting and re-initializing is really needed.

Reply:
As said earlier, sched_yield() will not make any difference when the child thread has already terminated. To avoid confusion, I have updated the code with comments. Also since I am checking all success and failure conditions, I would require 2 threads and hence initializing the attributes. Please have a look at latest patch. Thanks.